src/System.Private.CoreLib/shared/System/Buffers/ConfigurableArrayPool.cs

```diff
@@ -250,9 +252,12 @@ internal void Return(T[] array)
                 {
                     _lock.Enter(ref lockTaken);
 
-                    if (_index != 0)
+                    int index = _index;
+                    if (index != 0)
                     {
-                        _buffers[--_index] = array;
+                        index--;
+                        _buffers[index] = array;
+                        _index--;
```

Comment:
Are you sure these changes make any codegen difference?

Reply:
I believe yes, **very** slightly.

For the Return this is the diff:
```asm
    // before
    L006f: mov ecx, [ebp-0x2c]
    L0072: cmp dword [ecx+0xc], 0x0
    L0076: jz L0091
    L0078: mov ecx, [ebp-0x2c]
    L007b: mov ecx, [ecx+0x4]
    L007e: mov edx, [ebp-0x2c]
    L0081: mov edx, [edx+0xc]
    L0084: dec edx
    L0085: mov eax, [ebp-0x2c]
    L0088: mov [eax+0xc], edx
    L008b: push esi
    L008c: call 0x71d36700

    // now
    L006f: mov ecx, [ebp-0x2c]
    L0072: mov edx, [ecx+0xc]
    L0075: test edx, edx
    L0077: jz L008c
    L0079: dec edx
    L007a: mov ecx, [ebp-0x2c]
    L007d: mov ecx, [ecx+0x4]
    L0080: push esi
    L0081: call 0x71d36700
    L0086: mov ecx, [ebp-0x2c]
    L0089: dec dword [ecx+0xc]
```

@stephentoub [also asked about `_index = index;`](https://github.com/dotnet/corefx/pull/41829#discussion_r335563284) at the end instead of another decrement. The difference is two `mov`s vs one `dec`. I don't know enough asm to say which is better, I naively went with the size win.

Reply:
What are you using to see the resulting assembly?

Also, does it make any measurable difference when using the APIs?  If it's an almost imperceptible difference in codegen, compared to all the expenses (e.g. locks) in accessing these APIs, I'd assume the answer is "no", in which case we'd generally prefer readability/maintainability of the code.

Reply:
> What are you using to see the resulting assembly?

Just [SharpLab](https://sharplab.io/#v2:C4LghgzgtgPgAgBgARwIwDoAqALATgUzABMBLAOwHMBuAWACh7ElNMBPAB3yQF4UMBJMsADMAJloM65YPlxkwAGyQRCC/ERSikAIQCuAYwDW+YPQDe9JFaTTZ8pQWIB7MgtY2hSAPoAjXQDN/WQAZfEpgbAlrJHZcEgA3MBkkRyIXN2Y2TgBtAF087z9A2QgJS2tYhKSuAGV2cmCnI28FJsMoirjE5OlvciJ8AA8Oq3KrWzlFHQNjYAAKXqKg3FDw7AAaD2AkMl0oH1kAeX89YtwIAEoxpAs6aOivVubeMnwAdyQ6hra5i5H73wBZYQHg7d6ZDj4bK7fZHE5Akr5XL/B5LEJhCgRUFolYYiL/AC+12uE3sEJyuSQACUwvMrndrLd7tYWJC8gUcSDeICzqVrtFWRSkDjQbsFAoUdZ+dYfE4nEonoZMGBjGRQf5FCpNoontVTst1Zr8JKrMBcKxpVYmcyAYr0ABRISyOYEfxIRXK1V/eiW+4kN1zLz9IZIAA8woR53Qq0x2HpNvu1oTzJFvE52SDZAGg2RvoT6cz2YA1EXKS9dOKTcmdU09ZHsfXuOXK3nrESGcz2wn/OQdRaO4nW+MAx6VWELi02g7BiR5hqFCpvQPol2bb7/Ug5jX9HWzvGbUnk6mwR9BVCecsYxFc8u2z7b1Y4AB2CNnQkkp2TJRngo0oSiX5rkPAUsihSljzFCVfV9WV5XdNpPTCQ0F3wbVxVrGR9VkZCVCrM1+wTYCbR/JFX2BUELxKKsHjtR0ZFwF18DdUcvWoqUH2TXpg0GCjuLY5kNzmOZdGkC5uLDTcRKEC5OWjPE4yHQcOOTGVGzIkpsm4m8VJU9MtNFCsoOUnTCyGEt+OrdCd0wtSRRIEFIIsldfVXZke3sNxfSIgSRwQscyAnR4p3tGc5yNJcE1c+51wDbdd2WfdmW8+4IPBH9KNxNZtJtdtfWfdTcHfAcP3osk4AAFmpExdDkOYSKQMBcFwMBWESm4YLlBU/NVHDjV9fCvMUoKjAdT9GOY7rxzKYzh03UyeIAQl4BA2sIxTUUjCBsgAWm2+aywapqWosqLnI49y+0GmabF8oxEICycRpC2c5nnRcqyiqLSSmCqquAGqyAA+rGua1qgI6uCWKQ3g3r6jiBo45KaOCsbXXgu7/IinTFK4rMQ25PicYDcSlqQFbFKRm1uN2pzrAyrb9N4EHjvWunqe2k6XN9C7xQIg8ic3KGHuGwxpxe2Gsc7a52wJIA).

> Also, does it make any measurable difference when using the APIs?

I didn't run any benchmarks on this, just played around with codegen when basing an implementation on this code.

The difference in `Return` is 4 bytes under the SpinLock, I assume the real-world difference is imperceptible as you said.

For `Rent` the whole diff is 38 bytes (17 for moving the `_buckets` access into the try/finally and 21 for the index/bound check change), I would assume that could make a diff for time taken under the lock.

Reply:
Sharplab can't currently be used to check JIT ASM for .NET Core due to https://github.com/ashmind/SharpLab/issues/425